// @ionic/cli-utils/dist/guards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOG_LEVELS = ['debug', 'info', 'ok', 'warn', 'error'];
function isCommand(cmd) {
    return typeof cmd.run === 'function';
}
exports.isCommand = isCommand;
function isCommandPreRun(cmd) {
    return typeof cmd.preRun === 'function';
}
exports.isCommandPreRun = isCommandPreRun;
function isCommandPreInputsPrompt(cmd) {
    return typeof cmd.preInputsPrompt === 'function';
}
exports.isCommandPreInputsPrompt = isCommandPreInputsPrompt;
function isLogLevel(l) {
    const loglevel = l;
    return exports.LOG_LEVELS.includes(loglevel);
}
exports.isLogLevel = isLogLevel;
function isPackageJson(o) {
    const obj = o;
    return obj
        && typeof obj.version === 'string'
        && typeof obj.name === 'string'
        && typeof obj.dependencies === 'object'
        && typeof obj.devDependencies === 'object';
}
exports.isPackageJson = isPackageJson;
function isBowerJson(o) {
    const obj = o;
    return obj && typeof obj.name === 'string';
}
exports.isBowerJson = isBowerJson;
function isValidationErrorArray(e) {
    const err = e;
    return err && err[0]
        && typeof err[0].message === 'string'
        && typeof err[0].inputName === 'string';
}
exports.isValidationErrorArray = isValidationErrorArray;
function isSuperAgentError(e) {
    const err = e;
    return e && err.response && typeof err.response === 'object';
}
exports.isSuperAgentError = isSuperAgentError;
function isAPIResponseSuccess(r) {
    const res = r;
    return res && (typeof res.data === 'object' || typeof res.data === 'string');
}
exports.isAPIResponseSuccess = isAPIResponseSuccess;
function isAPIResponseError(r) {
    const res = r;
    return res && typeof res.error === 'object';
}
exports.isAPIResponseError = isAPIResponseError;
function isAppDetails(d) {
    const details = d;
    return details && typeof details === 'object'
        && typeof details.id === 'string'
        && typeof details.name === 'string'
        && typeof details.slug === 'string';
}
exports.isAppDetails = isAppDetails;
function isAppResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res) && isAppDetails(res.data);
}
exports.isAppResponse = isAppResponse;
function isAppsResponse(r) {
    let res = r;
    if (!isAPIResponseSuccess(res) || !Array.isArray(res.data)) {
        return false;
    }
    if (res.data.length > 0) {
        return isAppDetails(res.data[0]);
    }
    return true;
}
exports.isAppsResponse = isAppsResponse;
function isAuthTokensResponse(r) {
    const res = r;
    if (!isAPIResponseSuccess(res) || !Array.isArray(res.data)) {
        return false;
    }
    if (res.data.length > 0) {
        return typeof res.data[0].token === 'string'
            && typeof res.data[0].details === 'object'
            && typeof res.data[0].details.app_id === 'string'
            && typeof res.data[0].details.type === 'string'
            && typeof res.data[0].details.user_id === 'string';
    }
    return true;
}
exports.isAuthTokensResponse = isAuthTokensResponse;
function isLoginResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res) && typeof res.data.token === 'string';
}
exports.isLoginResponse = isLoginResponse;
function isDeployResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res)
        && typeof res.data.uuid === 'string'
        && typeof res.data.snapshot === 'string'
        && typeof res.data.channel === 'string';
}
exports.isDeployResponse = isDeployResponse;
function isDeployChannelResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res)
        && typeof res.data.uuid === 'string'
        && typeof res.data.tag === 'string';
}
exports.isDeployChannelResponse = isDeployChannelResponse;
function isDeploySnapshotResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res)
        && typeof res.data.uuid === 'string'
        && typeof res.data.url === 'string';
}
exports.isDeploySnapshotResponse = isDeploySnapshotResponse;
function isDeploySnapshotRequestResponse(r) {
    const res = r;
    return isAPIResponseSuccess(res)
        && typeof res.data.uuid === 'string'
        && typeof res.data.presigned_post === 'object'
        && typeof res.data.presigned_post.url === 'string'
        && res.data.presigned_post.fields && typeof res.data.presigned_post.fields === 'object';
}
exports.isDeploySnapshotRequestResponse = isDeploySnapshotRequestResponse;
