// @ionic/cli-utils/dist/lib/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util = require("util");
const chalk = require("chalk");
const guards_1 = require("../../guards");
const errors_1 = require("../errors");
const format_1 = require("./format");
class Logger {
    constructor(opts = {}) {
        this.level = opts.level || 'info';
        this.prefix = opts.prefix || '';
    }
    get level() {
        return this._level;
    }
    set level(v) {
        const s = v.toLowerCase();
        if (!guards_1.isLogLevel(s)) {
            throw new errors_1.FatalException(`Invalid log level '${chalk.bold(v)}' (choose from: ${guards_1.LOG_LEVELS.map(l => chalk.bold(l)).join(', ')})`);
        }
        this._level = s;
    }
    debug(...args) {
        this.log('debug', ...args);
    }
    info(...args) {
        this.log('info', ...args);
    }
    ok(...args) {
        this.log('ok', ...args);
    }
    warn(...args) {
        this.log('warn', ...args);
    }
    error(...args) {
        this.log('error', ...args);
    }
    msg(...args) {
        console.log.apply(console, args);
    }
    nl(num = 1) {
        console.log('\n'.repeat(num - 1));
    }
    shouldLog(level) {
        return guards_1.LOG_LEVELS.indexOf(level) >= guards_1.LOG_LEVELS.indexOf(this.level);
    }
    log(level, ...args) {
        if (this.shouldLog(level)) {
            let prefix = this.prefix;
            if (prefix) {
                args[0] = util.format(prefix, args[0]);
            }
            if (['ok', 'warn', 'error'].includes(level)) {
                for (let [i, arg] of args.entries()) {
                    if (typeof arg === 'string') {
                        args[i] = arg.split('\n').map((l, i) => i > 0 ? `${format_1.indent(level.length + 2)} ${l}` : l).join('\n');
                    }
                }
            }
            const status = chalk.bold.bgBlack;
            const b = chalk.dim;
            switch (level) {
                case 'debug':
                case 'info':
                    console.info(util.format.apply(util, args));
                    break;
                case 'ok':
                    console.info(util.format.apply(util, [b('[') + status.green('OK') + b(']'), ...args]));
                    break;
                case 'warn':
                    console.warn(util.format.apply(util, [b('[') + status.yellow('WARN') + b(']'), ...args]));
                    break;
                case 'error':
                    console.error(util.format.apply(util, [b('[') + status.red('ERROR') + b(']'), ...args]));
                    break;
            }
        }
    }
}
exports.Logger = Logger;
