// @ionic/cli-utils/dist/lib/utils/task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tty = require("tty");
const chalk = require("chalk");
const format_1 = require("./format");
const modules_1 = require("../modules");
const FRAMES = process.platform === 'win32' ?
    ['-', '\\', '|', '/'] :
    ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
exports.TASKS = [];
class Spinner {
    constructor(frames = FRAMES) {
        this.frames = frames;
        this.i = 0;
    }
    frame() {
        return this.frames[this.i = ++this.i % this.frames.length];
    }
}
class Task {
    constructor(msg) {
        this.msg = msg;
        this.running = false;
        this.spinner = new Spinner();
        const inquirer = modules_1.load('inquirer');
        this.bottomBar = new inquirer.ui.BottomBar();
        exports.TASKS.push(this);
    }
    start() {
        if (!this.running) {
            this.intervalId = setInterval(() => { this.tick(); }, 50);
        }
        this.running = true;
        return this;
    }
    tick() {
        this.bottomBar.updateBottomBar(this.format());
        return this;
    }
    progress(prog, total) {
        if (this.running) {
            if (!this.progressBar) {
                const term = tty;
                const ProgressBar = modules_1.load('progress');
                const s = new term.WriteStream();
                s.columns = s.columns || 80;
                this.progressBar = new ProgressBar('[:bar] :percent', {
                    total: total,
                    width: 30,
                    stream: s,
                });
            }
            const progbar = this.progressBar;
            progbar.curr = prog;
            if (prog < total) {
                this.progressBar.tick(0);
            }
            this.tick();
        }
        return this;
    }
    format() {
        const progbar = this.progressBar;
        const progress = progbar ? progbar.lastDraw.trim() : '';
        return `${chalk.bold(this.spinner.frame())} ${this.msg} ${progress}`;
    }
    clear() {
        clearInterval(this.intervalId);
        this.bottomBar.updateBottomBar('');
        this.bottomBar.close();
        return this;
    }
    end() {
        this.running = false;
        this.tick();
        this.clear();
        return this;
    }
    succeed() {
        if (this.running) {
            this.end();
            console.log(`${chalk.green(format_1.ICON_SUCCESS_GREEN)} ${this.msg} - done!`);
        }
        return this;
    }
    fail() {
        if (this.running) {
            this.end();
            console.error(`${chalk.red(format_1.ICON_FAILURE_RED)} ${this.msg} - failed!`);
        }
        return this;
    }
}
exports.Task = Task;
class TaskChain {
    next(msg) {
        if (this.currentTask) {
            this.currentTask.succeed();
        }
        this.currentTask = new Task(msg);
        this.currentTask.start();
        return this.currentTask;
    }
    updateMsg(msg) {
        if (this.currentTask) {
            this.currentTask.msg = msg;
        }
        return this;
    }
    end() {
        if (this.currentTask) {
            this.currentTask.succeed();
            this.currentTask = undefined;
        }
        return this;
    }
    fail() {
        if (this.currentTask) {
            this.currentTask.fail();
        }
        return this;
    }
}
exports.TaskChain = TaskChain;
